chart theming, url backed state & universal zoom

gozart1 opened the pull request:
Changes include:

- Use [History pushState](https://developer.mozilla.org/en-US/docs/Web/API/History_API) to add sharable urls for each chart
- Night mode chart theme
- Use color palette from https://github.com/decred/dcrdata/issues/451 for series colors
- Utilize all available horizontal screen real estate
- Enable client side zoom helper for all charts
- Rename "Tickets By Output" to "Ticket Purchase"
- Remove grouping helper for "Ticket Purchase" chart and instead show them as separate options: "Ticket Purchases By Block" and "Ticket Purchases By Price Window" ( simplifies ui & url scheme )
- Move hover state into external div instead of overlaying it on the chart ( this way we get a permanently visible legend, which is useful for screenshots and touch interfaces )

I'd like to further improve the url backing by including the selected range or zoom. But I think we should consider hooking that into the backend. That way we can set a default zoom such as one month and only load the full range if a users actually wants that. This will make the charts load a whole lot faster. In any case, url backed state for zoom/selection should be separate PR since it probably warrants further discussion and there is no need for it to be a blocker to this PR.

<img width="1274" alt="Day" src="https://user-images.githubusercontent.com/25571523/43821970-83d03228-9a9f-11e8-8e8c-f8e4bf2c66bf.png">
<img width="1275" alt="Night" src="https://user-images.githubusercontent.com/25571523/43821977-867bf962-9a9f-11e8-90a8-e428f6f2f5ca.png">

linnutee commented:
Can consider aligning "Zoom: Day, Week, Month" with the rest of the controls UI style/colors, so it wouldn't be competing with the chart elements.

chappjc commented:
For the Gopkg.lock conflicts, there should be 5 `revision` changes that you'll want to take, and you can ignore all the `digest =` lines.  But after the resolve, and before commit of the rebase, I'd update dep (go get **-u** -v github.com/golang/dep/cmd/dep) and run a `dep ensure` to get it all spiffy.

chappjc commented:
I think this is pretty much needed for 3.0, so slated as such.

xaur commented:
Nice colors!

Ideas:

* Lay out "Window/Solo/Pooled" horizontal, in same lane as other controls, when there's enough width.
* When not enough width or if you decide to keep this block vertical, at least right-align the numbers.
* Add ticket price data points and second Y axis for that on the right.

gozart1 commented:
Thanks for all the feedback, very helpful!
Here's another rev on the chart controls and legend that seems well to work under a wide variety of circumstances ( labels + values + screen sizes ).

Added a control for applying a moving average to the dataset. The rollups are more interesting for some charts vs others. So, I'm not sure if we should always show it, but I don't see why not at this point since there aren't that many controls competing for attention. Also, since the computation is done client side, it can get a bit heavy when computing a large moving average for large datasets ( I've set a max moving average period to 10,000, for the sake of putting some sort of upper limit to prevent completely freezing up the browser tab ).

<img width="1274" alt="tickets-month" src="https://user-images.githubusercontent.com/25571523/44026326-8b920c00-9ea7-11e8-911a-10f22e59acdf.png">
<img width="1277" alt="tickets-raw" src="https://user-images.githubusercontent.com/25571523/44026328-8cec9278-9ea7-11e8-98b0-cdcdc08dfc69.png">

xaur commented:
Is there a reason why you changed the zoom buttons sorting from ascending to descending scale?

gozart1 commented:
@xaur Since 'All' is the default it just seems more natural to me for the default to be the first option. 
The reason 'All' is the default is because the current implementation loads all the entire dataset, so we might as well show it after putting the user through the hard work of loading and parsing it.

As mentioned above, I think we should probably switch to server-side zoom in the future, and default to only loading a recent portion of the dataset (or some sort of aggregate) for the sake of performance. 

But the main reason to go from bigger to smaller is because that lines up with how the range slider will update. The further right an option is on the zoom helper, the further to the right the start of the selected range will be. Again, subtle, but it just feels more natural to me for those controls to be aligned vs inverted.

xaur commented:
@gozart1 thanks for sharing. I noticed it because it stands out compared to most charts I used to see (coinmarketcap, onchainfx, etc) that all have their buttons go from smaller to bigger scale left to right. But your considerations are reasonable and interesting. No hard opinion here, curious to see how it goes!

gozart1 commented:
Good point. Now that I look around they really do all go from smaller to bigger. But those are all trading sites, it must be a conspiracy to bias people towards thinking along short term time horizons! j/k

But in all seriousness, there likely is a very good reason it's so ubiquitous. Maybe it is something along the lines of this: If a user is reaching for the zoomer, then prioritizing the option which is _not_ zooming ('All') is less ergonomic. Genuinely curious now how this came about.

I'm happy to flip it back to ascending. I didn't realize this was such a convention and think flouting UX convention should be avoided unless there is a strong case.

xaur commented:
Haha, long term by default. I like it.

chappjc commented:
Been testing and I love it.  But after the mega reorg, we clearly need to limit the queries to mainchain only (TODO in future backend PR).  From the transactions per block chart:

![image](https://user-images.githubusercontent.com/9373513/44169002-b8297d00-a098-11e8-8b82-31f60f2ef74e.png)

chappjc commented:
The ticket purchases by price window could use some date or block number info on the x axis too, but if that requires API changes it can wait too.

chappjc commented:
To reiterate my discussion with Migwi, we'll have to reconsider how we present the Tickets By Output Count as it's really impossible to say pool or solo based on number of outputs.  The best I think we can do without being misleading is to simply show 3 output / 5 output / other.

chappjc commented:
BTW, I'm really happy with the response times.  The backend served every API endpoint in <1ms each.

```"GET http://127.0.0.1:8777/api/chart/tx-per-day HTTP/1.1" from 127.0.0.1:52600 - 200 275B in 35.687µs
"GET http://127.0.0.1:8777/api/chart/tx-per-block HTTP/1.1" from 127.0.0.1:52622 - 200 43393B in 367.822µs
"GET http://127.0.0.1:8777/api/chart/pow-difficulty HTTP/1.1" from 127.0.0.1:52622 - 200 1035B in 35.849µs
"GET http://127.0.0.1:8777/api/chart/coin-supply HTTP/1.1" from 127.0.0.1:52654 - 200 131507B in 954.794µs
"GET http://127.0.0.1:8777/api/chart/fee-per-block HTTP/1.1" from 127.0.0.1:52686 - 200 95170B in 818.505µs
"GET http://127.0.0.1:8777/api/chart/coin-supply HTTP/1.1" from 127.0.0.1:52708 - 200 131507B in 909.299µs
"GET http://127.0.0.1:8777/api/chart/fee-per-block HTTP/1.1" from 127.0.0.1:52730 - 200 95170B in 875.099µs
"GET http://127.0.0.1:8777/api/chart/duration-btw-blocks HTTP/1.1" from 127.0.0.1:52730 - 200 47602B in 598.113µs
"GET http://127.0.0.1:8777/api/chart/ticket-spend-type HTTP/1.1" from 127.0.0.1:52764 - 200 27355B in 312.142µs
"GET http://127.0.0.1:8777/api/chart/ticket-by-outputs-windows HTTP/1.1" from 127.0.0.1:52854 - 200 307B in 35.118µs
"GET http://127.0.0.1:8777/api/chart/ticket-by-outputs-blocks HTTP/1.1" from 127.0.0.1:53068 - 200 28313B in 446.388µs
```

This should be reasonable for the public server to handle.

chappjc commented on public/js/util.js:
```diff
@@ -0,0 +1,8 @@
+// a place for handy functions utility that can be shared accross the client side application
```
across

chappjc commented on views/charts.tmpl:
```diff
@@ -5,69 +5,142 @@
+            <div class="spinner-wrapper">
+                <div class="spinner-centerer d-flex align-items-center justify-content-center">
+                    <div class="spinner">
+                        <i></i><i></i><i></i><i></i><i></i><i></i><i></i><i></i><i></i><i></i><i></i><i></i>
```
What do all these `<i>`s do?

gozart1 replied:
They are the spokes of the spinner wheel
<img width="637" alt="spinner" src="https://user-images.githubusercontent.com/25571523/44196915-1d5d8c80-a0f2-11e8-9376-6ccdce3ba928.png">

gozart1 commented:
I've added the block range of price window since we can just compute it on the client.
Changed the Ticket Purchase charts back to Ticket Output Counts. What do you think of the labels suggesting the inference "(likely Solo)" ?
<img width="1275" alt="likely" src="https://user-images.githubusercontent.com/25571523/44199526-3cabe800-a0f9-11e8-9a2b-88157e88211a.png">

gozart1 commented:
Other than settling on a nomenclature for the Ticket Outputs charts, I think this PR is almost ready to go. 

One last item is either to remove the query param parser lib I've added here https://github.com/decred/dcrdata/pull/598/files#diff-5bab677d36ae3694a8a856712a6b0e81, or put it to use. What I had in my is to represent all the chart controls via history pushState in the url, so that an exact view is sharable via a link.
 
The proposed structure is:
 ```/charts#?type=${chart_type}&start=${start}&end=${end}&rollPeriod=${rollPeriod}```

So if blocks 1 - 12,356 of Coin Supply is selected the url would be:
```/charts#?type=coin-supply&start=1&end=12356```

When a chart initially loads, or if a specific zoom is clicked the url would look like this
```/charts#?type=coin-supply&zoom=0```
So a url would either have zoom or start/end, never both.

Another thing to note is in this design, everything before the `#` is considered server-side param, everything after it is a purely the concern of the client side.

It's a bit complicated, so I backed out on implementing it for now. 
If we implement this here, the end user would expect it to work like that in other places as well. So I want to make sure everyone is on board before getting too deep.

I'm leaning towards omitting url backed state for now, and revisit this one we have more views with advanced client side state as that may surface some additional considerations that might help us decide on things like wether to bring in a dedicated library for client side routing.
 
@dmigwi Any thoughts on this since you are in the thick of it with your charts work?

gozart1 commented:
OK, have removed the unused query param parsing lib. Let's merge this and hash out an appropriate client side routing architecture with a separate Issue/PR.
